libdnf/system/state.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/system/state.hpp"
+
+#include "libdnf/utils/fs.hpp"
+
+#include <toml.hpp>
+#include <fstream>
+
+namespace libdnf::system {
+
+State::State(
+    const std::filesystem::path & installroot,
+    const std::filesystem::path & dir_path)
+    : path(installroot / dir_path) {
+    load();
+}
+
+
+libdnf::transaction::TransactionItemReason State::get_reason(const std::string & na) {
+    auto it = reasons.find(na);
+    if (it == reasons.end()) {
+        return libdnf::transaction::TransactionItemReason::DEPENDENCY;
+    }
+
+    return it->second;
+}
+
+
+void State::set_reason(const std::string & na, libdnf::transaction::TransactionItemReason reason) {
+    if (reason == libdnf::transaction::TransactionItemReason::DEPENDENCY) {
```

Comment:
I see that the design is not extendable for other kind of reasons. Like weak dependencies are automatically degraded to dependencies. In DNF we fight with requests when users asked why weak dependencies appeared in transaction or if the dependency is dependency of userinstalled package or of a weak dependent package. With DNF5 we promote the problem to state when user will be unable to get even information if the package was installed as weak dependency or not.

Reply:
Yes, as from this point of view, it doesn't matter if the dependency is weak.

This doesn't mean we can't store the fine-grained reason in transaction history db, so the user can always find out why the package was originally installed. However, storing this reason (i.e. WEAK_DEPENDENCY) in the system state is actually not reliable, as it can change in subsequent transactions in which the weak dependency isn't present at all. Example: Package A has a weak dep on package X, the user installs A and X gets pulled in. Later the user installs B, which has a regular dependency on X. Now the reason for X is regular dependency, even though the stored reason for it would be WEAK_DEPENDENCY.